File: 2021/python/day11.py
def load_file():
    with open('../inputs/input11') as file:
        return file.read()


def create_energy_map():
    data = load_file()
    return [[int(c) for c in line] for line in data.splitlines()]


def neighbors(x, y, width, height):
    n = [(x - 1, y - 1), (x, y - 1), (x + 1, y - 1), (x - 1, y), (x, y),
         (x + 1, y), (x - 1, y + 1), (x, y + 1), (x + 1, y + 1)]
    return list(filter(lambda t: 0 <= t[0] < width and 0 <= t[1] < height, n))


def update(positions, energy_map, flashed):
    flash = []
    for x, y in positions:
        if (x, y) not in flashed:
            energy_map[y][x] += 1
            if energy_map[y][x] > 9:
                flash.append((x, y))
                energy_map[y][x] = 0
    return flash


def sum_is_zero(energy_map):
    for y in range(len(energy_map)):
        for x in range(len(energy_map[0])):
            if energy_map[y][x] > 0:
                return False
    return True


def run(steps, energy_map):
    width = len(energy_map[0])
    height = len(energy_map)
    positions = []
    for y in range(height):
        for x in range(width):
            positions.append((x, y))

    flash_counter = 0
    step_zero = None
    for step in range(steps):
        if sum_is_zero(energy_map):
            return None, step
        current_flashes = update(positions, energy_map, [])
        skip_flashes = set(current_flashes)
        flash_counter += len(current_flashes)
        while len(current_flashes) > 0:
            x, y = current_flashes.pop(0)
            n_flashes = update(neighbors(x, y, width, height), energy_map,
                               skip_flashes)
            flash_counter += len(n_flashes)
            skip_flashes.update(n_flashes)
            current_flashes.extend(n_flashes)

    return flash_counter, step_zero


def part1():
    return run(100, create_energy_map())[0]


def part2():
    return run(1000, create_energy_map())[1]


print(part1())
print(part2())

File: 2021/rust/src/day11.rs
use std::collections::HashSet;

fn create_energy_level(lines: &Vec<String>) -> Vec<Vec<usize>> {
    lines
        .iter()
        .map(|line| {
            line.chars()
                .map(|c| c.to_digit(10).unwrap() as usize)
                .collect()
        })
        .collect()
}

fn get_neighbors(x: usize, y: usize, width: usize, height: usize) -> Vec<(usize, usize)> {
    let mut neighbors = vec![(x, y + 1), (x + 1, y), (x + 1, y + 1)];
    if x >= 1 {
        neighbors.push((x - 1, y));
        neighbors.push((x - 1, y + 1));
    }
    if x >= 1 && y >= 1 {
        neighbors.push((x - 1, y - 1));
    }
    if y >= 1 {
        neighbors.push((x, y - 1));
        neighbors.push((x + 1, y - 1));
    }
    neighbors
        .into_iter()
        .filter(|n| n.0 < width && n.1 < height)
        .collect()
}

fn update(
    positions: &Vec<(usize, usize)>,
    energy_level: &mut Vec<Vec<usize>>,
    flashed: &HashSet<(usize, usize)>,
) -> Vec<(usize, usize)> {
    let mut flashing = vec![];
    for position in positions {
        if flashed.contains(position) {
            continue;
        }
        let (x, y) = position;
        energy_level[*y][*x] += 1;
        if energy_level[*y][*x] > 9 {
            flashing.push((*x, *y));
            energy_level[*y][*x] = 0;
        }
    }
    flashing
}

fn is_sum_zero(energy_level: &Vec<Vec<usize>>) -> bool {
    let height = energy_level.len();
    let width = energy_level[0].len();
    for y in 0..height {
        for x in 0..width {
            if energy_level[y][x] > 0 {
                return false;
            }
        }
    }
    true
}

fn run(steps: usize, energy_level: &mut Vec<Vec<usize>>) -> usize {
    let height = energy_level.len();
    let width = energy_level[0].len();
    let mut positions = vec![];
    for y in 0..height {
        for x in 0..width {
            positions.push((x, y));
        }
    }
    let mut flash_counter = 0;
    for step in 0..steps {
        if is_sum_zero(&energy_level) {
            return step;
        }
        let mut current_flashes =
            update(&positions, energy_level, &HashSet::<(usize, usize)>::new());

        let mut skip_flashes = HashSet::<(usize, usize)>::from_iter(current_flashes.clone());
        flash_counter += current_flashes.len();
        while !current_flashes.is_empty() {
            let (x, y) = current_flashes.pop().unwrap();
            let mut neighbor_flashes = update(
                &get_neighbors(x, y, width, height),
                energy_level,
                &skip_flashes,
            );
            flash_counter += neighbor_flashes.len();
            skip_flashes.extend(neighbor_flashes.clone());
            current_flashes.append(&mut neighbor_flashes);
        }
    }
    flash_counter
}

pub fn part1(lines: &Vec<String>) -> usize {
    run(100, &mut create_energy_level(lines))
}

pub fn part2(lines: &Vec<String>) -> usize {
    run(1000, &mut create_energy_level(lines))
}
